File: python/string.py
# STRINGS
# For string you can use ' or "
'a'
"b"
# For multiline strings use ''' or """
'''aaa
aaa''' # 'aaa\naaa'
"""bbb
bbb""" # "bbb\nbbb"

a = 'aaa' 'bbb' # literals go together is joined. (as there is +)
print(a) # 'aaabbb'
# You can write long line like this
f('foo '
  'bar '
  'boo') # 'foo bar boo'
# For defining variable
x = ('foo '
     'bar '
     'boo')
     
# Special symbols
# \' - '
# \" - "
# \t - tab
# \n - new line
# \xhh - hex symbol hh
# \\ - \

# You can use raw string to python understand strings as is:
print('\tHello') # '	Hello'
print(r'\tHello') # '\tHello'



# We should code symbols somehow
# At first lets match symbol to number
# ASCII - table that matches 127 symbols
# Unicode - table that matches 120000 symbols
# Now we should convert Unicode number to bytes
# There is several ways
# UTF-8 - uses 1-4 8 bit code units
# UTF-16 - uses 1 or 2 16 bit code units
# UTF-32 - uses 1 32bit code unit
# USC-1, USC-2, USC-4



# It uses unicode in python. We can use any characters from it. (English, Japaniese, Russian)
# Python uses USC-1, USC-2, USC-4 depending on the character range in a string
ord('a') # get unicode code the symbol 'a'
chr(85) # get character by unicode code

# You can add unicode symbol with its codes. Use \n and then code
print('\u0072') # \u for characters that can encode with 2 bytes  
print('\U0001F921') # \U for 4 byte characters
print('sometext\U0001F921\u0072anothertext') # you can use them in a string
print('\N{Japanese Goblin}') # You can use name for character



# STRING METHODS
# Change case
print(s1.upper()) # 'HELLO WORLD!'
print(s1.lower()) # 'hello world!'
print(s1.capitalize()) # 'Hello world!'
print(s1.title()) # 'Hello World!'
print(s1.title().swapcase()) # 'hELLO wORLD!'

# Alignment
print('foo bar'.ljust(20, '-')) # foo bar-------------
print('foo bar'.rjust(20, '-')) # -------------foo bar
print('foo bar'.center(20, '-')) # ------foo bar-------
print('foo bar'.center(20)) # '      foo bar       '

# Strips
# Strips remove characters in the left, right or both side which in an argument string
print(']>>foo bar<<['.lstrip(']>')) # 'foo bar<<['
print(']>>foo bar<<['.rstrip('<[')) # ']>>foo bar'
print(']>>foo bar<<['.strip('><[]')) # 'foo bar'
print('\n\t some text \t \n'.strip()) # 'some text'

# Split / rsplit
print('some    text'.split(' ')) # ['some', '', '', '', 'text'] - if to separators near - add empty string
print(' \t  some  \n   text   '.split()) # ['some', 'text'] - if there is no arguments, it joins all near space characters and then split
# It works only without argument
# You can specify number of splittings
print('foo.bar.bar.boo'.split('.', 2)) # ['foo', 'bar', 'bar.boo']
print('foo.bar.baz.boo'.rsplit('.', 2)) # ['foo.bar', 'baz', 'boo']

# Partition
# Always makes one split, and returns tuple with 3 values
print('foo.bar.bar.boo'.partition('.')) # ('foo', '.', 'bar.bar.boo')
print('foo.bar.bar.boo'.rpartition('.')) # ('foo.bar.bar', '.', 'boo')
print('foobar'.partition('.')) # ('foobar', '', '')

# Join
print(', '.join(['a', 'b', 'c'])) # 'a, b, c'
# Use join instead of constance string concatination in a for loop
# But cpython optimize this scenario, time will be almost the same

# Contanins, startswith, endswith
print('aca' in 'abacaba') # True
print('ada' not in 'abacaba') # True
print('abacaba'.startswith('aba')) # True
print('abacaba'.endswith('caba')) # True

# Find, index
# Index throws exceptions, if can't find substring, find return -1
print('abaracadabra'.find('ara')) # 2
print('abaracadabra'.find('ara', 2, 10)) # -1 - not this substring
print('abaracadabra'.index('ara')) # 2
print('abaracadabra'.index('ara', 2, 10)) # error
print('abacaba'.rindex('aba')) # 4 find begining the end
print('abacaba'.rfind('aba')) # 4

# Replace, translate
print('abracadabra'.replace('br', '**')) # 'a**acada**a'
print('abracadabra'.replace('br', '**', 1)) # 'a**acadabra'
print('abracadabra'.translate({ ord('a'): '*', ord('b'): '*' })) # replace sevaral symbols. Key should be a key of a symbol

File: java/string.java
// STRINGS
// String aren't modified
// CREATION
String s1 = "Hello world!";
String s2 = new String("Hello World!");
String s3 = new String(new char[] {
		'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd', '!'
});



// METHODS
"Hello World!".startWith("Hello");
"Hello World!".endsWith("World");
"Hello World!".contains(" ");
"Hello World!".equalsIgnoreCase("HeLLo WoRLd!");
"Hello World!".indexOf('e');
"Hello World!".indexOf("oo");
"Hello World!".lastIndexOf("o", 10); // search from end, you can specify
                                     // index from which to search
"Hello World!".replace('o', 'a'); // replace all occurances
"Hello World!".replace("Hello", "Goodbye");
"Hello World!".replaceFirst("^Hello", "Goodbye"); // replace one occurances
                                                  // use regex to specify string
"Hello World!".replaceAll("\\w", "x"); // replace all occurances, use regex to specify string
"   Hello World!   ".trim();
"Hello World!".substring(6);
"Hello World!".substring(6, 11);
"Hello World!".split(" ");



// STRING FORMATTING
// - Concatination: +
// - String.concat() - do not use it
// - String.format()
String make = "make";
String java = "java";
String great = "great";
String again = "again";
System.out.println(String.format("%s %s %s %s", make, java, great, again));
// You can specify locale
System.out.println(String.format(Locale.ENGLISH,"%s %s %s %s", make, java, great, again));

// - MessageFormat.format()
//   MessageFormat has placeholders that are different from String.format
String make = "make";
        String java = "java";
        String great = "great";
        String again = "again";
        System.out.println(MessageFormat.format("%s %s %s %s", make, java, great, again));

// - StringBuilder(StringBuffer)
StringBuilder builder = new StringBuilder();
for(Weekday weekday: Weekday.values()) {
	builder.append(weekday).append("\n");
}
System.out.println(builder.toString());
// - String.join()
// - StringJoiner - similar to StringBuilder
// - "*".repeat(100)



// Note, the methods toLowerCase, toUpperCase uses current computer locale.
// With english locale this works:
String string = "INT";
String lower = string.toLowerCase();
System.out.println(lower.equals("int")); //true
// But in turkish there are i with dot and without. So it is not works.
System.out.println(lower.equals("int")); // false
// To avoid this you can use the root locale
String string = "INT";
String lower = string.toLowerCase(Locale.ROOT);



// By default, java uses system encoding
// When you convert string to bytes and viseversa, better to specify encoding
System.out.println(new String(new byte[] {
                72, 101, 108, 108, 111, 32, 87, 111, 114, 108, 100, 33
        }, StandardCharsets.UTF_8));
System.out.println(Arrays.toString("Hello World!".getBytes(StandardCharsets.UTF_8)));

// The same is for reading string from files, we should specify how to convert
// file bytes to symbols
new FileReader("/etc/passwd", StandardCharsets.UTF_8);
